improver/nowcasting/utilities.py

```diff
@@ -103,6 +102,122 @@ def process(self, radar_data, coverage):
 
 
+class FillRadarHoles(BasePlugin):
+    """Interpolate small "no data" regions in the radar composite domain
+    """
+
+    @staticmethod
+    def _rr_to_log_rr(data):
+        """Convert a masked array of rainrates in mm/h into log space.  This
+        does not preserve values below 0.001; however, since the radar encodes
+        trace rain rates with a value of 0.03 mm/h, this should not have any
+        effect on "real" data.
+        """
+        min_rr_mmh = 0.001
+        result = np.where(data > min_rr_mmh, np.log10(data), np.nan)
+        return np.ma.MaskedArray(result, mask=data.mask)
+
+    @staticmethod
+    def _log_rr_to_rr(data):
+        """Convert a masked array of log rainrates into mm/h"""
+        result = np.where(np.isfinite(data), np.power(10, data), 0.0)
+        return np.ma.MaskedArray(result, mask=data.mask)
+
+    @staticmethod
+    def _find_and_interpolate_speckle(log_rr):
+        """Identify and interpolate "speckle" points.
+
+        The constants for speckle identification ("p_masked" and "r_speckle")
+        have been empirically tuned for UK radar data.  With the constants
+        as set, this method will flag "holes" of up to 24 pixels in size.
+        The neighbourhood size for interpolation has been chosen to match these
+        constants, ensuring that there will always be valid data in the
+        neighbourhood over which averaging is performed.
+
+        Args:
+            log_rr (numpy.ma.MaskedArray):
+                Masked array of rainrates in log10(mm/h)
+
+        Returns:
+            numpy.ma.MaskedArray:
+                Masked array of interpolated rainrates in log10(mm/h)
+        """
+        # minimum proportion of masked neighbours, below which a pixel is
+        # "speckle"
+        p_masked = 0.3
+        # radius of neighbourhood over which to search for masked neighbours
+        r_speckle = 4
+        # radius of neighbourhood from which to calculate interpolated values
+        r_interp = 2
+
+        interpolated_points = np.ma.MaskedArray(
+            np.full_like(log_rr, 0), mask=np.full_like(log_rr, True)
+        )
+
+        for y in range(r_speckle, log_rr.shape[0] - r_speckle - 1, 1):
+            for x in range(r_speckle, log_rr.shape[1] - r_speckle - 1, 1):
+                if log_rr.mask[y, x]:
+                    nbhood = np.mean(
+                        log_rr.mask[
+                            y - r_speckle : y + r_speckle + 1,
+                            x - r_speckle : x + r_speckle + 1,
+                        ]
+                    )
+                    if nbhood < p_masked:
+                        surroundings = log_rr[
+                            y - r_interp : y + r_interp + 1,
+                            x - r_interp : x + r_interp + 1,
+                        ]
+                        interpolated_points[y, x] = np.mean(
+                            surroundings[np.where(surroundings.mask == False)]
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Comparison to False should be 'not expr'](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5409292)